File: python/0222完全二叉树的节点个数.py
from binarytree import TreeNode

def countNodes1(root: TreeNode) -> int:
    count=0

    def _preTraversal(root):
        nonlocal count
        if(root):
            count+=1
            _preTraversal(root.left)
            _preTraversal(root.right)
    
    _preTraversal(root)
    return count

def countNodes2(root:TreeNode)->int:
    if(not root):
        return 0
    return 1+countNodes2(root.left)+countNodes2(root.right)

# 利用完全二叉树性质
def countNodes(root:TreeNode)->int:
    if(not root):
        return 0
    
    leftdepth=rightdepth=0

    p=root.left
    while(p):
        leftdepth+=1
        p=p.left
    p=root.right
    while(p):
        leftdepth+=1
        p=p.right
    
    # 左右子树深度相同则为满二叉树
    if(leftdepth==rightdepth):
        return (2<<leftdepth)-1
    
    # 若不为满二叉树则继续检测左右子树是否分别为满二叉树
    return 1+countNodes(root.left)+countNodes(root.right)

File: cpp/0222完全二叉树的节点个数.cpp
#include<deque>
#include"mybitree.h"

int countNodes(TreeNode *root)
{
    if(!root)
        return 0;
    deque<TreeNode *> nodes;
    nodes.push_back(root);
    int res = 0;

    while(!nodes.empty())
    {
        int size = nodes.size();
        
        for (int i = 0; i < size; ++i)
        {
            TreeNode *crt = nodes.front();
            nodes.pop_front();

            if(crt)
            {
                ++res;
                nodes.push_back(crt->left);
                nodes.push_back(crt->right);
            }
        }
    }

    return res;
}